src/mongo_logger.ts

```diff
@@ -122,58 +178,262 @@ function resolveLogPath(
   }
-): Writable {
+): Writable | MongoDBLogWritable {
   const isValidLogDestinationString = (destination: string) =>
     ['stdout', 'stderr'].includes(destination.toLowerCase());
   if (typeof mongodbLogPath === 'string' && isValidLogDestinationString(mongodbLogPath)) {
     return mongodbLogPath.toLowerCase() === 'stderr' ? process.stderr : process.stdout;
   }
 
-  // TODO(NODE-4813): check for minimal interface instead of instanceof Writable
-  if (typeof mongodbLogPath === 'object' && mongodbLogPath instanceof Writable) {
+  if (
+    mongodbLogPath &&
+    typeof mongodbLogPath === 'object' &&
+    (mongodbLogPath instanceof Writable || mongodbLogPath?.write)
+  ) {
     return mongodbLogPath;
   }
 
   if (typeof MONGODB_LOG_PATH === 'string' && isValidLogDestinationString(MONGODB_LOG_PATH)) {
     return MONGODB_LOG_PATH.toLowerCase() === 'stderr' ? process.stderr : process.stdout;
   }
 
-  return process.stderr;
+  return process.stderr as unknown as MongoDBLogWritable;
+}
+
+/** @internal */
+export interface Log extends Record<string, any> {
+  s: SeverityLevel;
+  t: Date;
+  c: MongoLoggableComponent;
+  message?: string;
+}
+
+/** @internal */
+export interface MongoDBLogWritable {
+  write(log: Log): unknown;
+}
+
+function compareSeverity(s0: SeverityLevel, s1: SeverityLevel): 1 | 0 | -1 {
+  const s0Num = SEVERITY_LEVEL_MAP.getNumericSeverityLevel(s0);
+  const s1Num = SEVERITY_LEVEL_MAP.getNumericSeverityLevel(s1);
+
+  return s0Num < s1Num ? -1 : s0Num > s1Num ? 1 : 0;
+}
+
+/** @internal */
+export interface Loggable extends Record<string, any> {
+  toLog?(): Record<string, any>;
+}
+
+function DEFAULT_LOG_TRANSFORM(logObject: Loggable): Omit<Log, 's' | 't' | 'c'> {
+  let log: Omit<Log, 's' | 't' | 'c'> = {};
+
+  const getHostPort = (address: string): { host: string; port: number } => {
+    const hostAddress = new HostAddress(address);
+
+    // NOTE: Should only default when the address is a socket address
+    const host = hostAddress.host ?? '';
+    const port = hostAddress.port ?? 0;
+    return { host, port };
+  };
+
+  const attachCommandFields = (
+    l: any,
+    ev: CommandStartedEvent | CommandSucceededEvent | CommandFailedEvent
+  ) => {
+    l.commandName = ev.commandName;
+    l.requestId = ev.requestId;
+    l.driverConnectionId = ev?.connectionId;
+    const { host, port } = getHostPort(ev.address);
+    l.serverHost = host;
+    l.serverPort = port;
+    if (ev?.serviceId) {
+      l.serviceId = ev.serviceId.toHexString();
+    }
+
+    return l;
+  };
+
+  const attachConnectionFields = (l: any, ev: ConnectionPoolMonitoringEvent) => {
+    const { host, port } = getHostPort(ev.address);
+    l.serverHost = host;
+    l.serverPort = port;
+
+    return l;
+  };
+
+  let ev;
+  if (APM_EVENTS.includes(logObject.name)) {
+    log = attachCommandFields(log, logObject as any);
+    switch (logObject.name) {
+      case COMMAND_STARTED:
+        ev = logObject as CommandStartedEvent;
```

Comment:
I'm hoping that if we can improve the type information of `name` we can remove this casting. I think you'll have to make union out of Loggable